connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -279,10 +284,61 @@ public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callb
+                    } else {
+                        log.debug("Successfully flushed tombstone offsets to secondary backing store");
+                    }
+                }
+            });
+            try {
+                // For EOS, there is no timeout for offset commit and it is allowed to take as much time as needed for
+                // commits. We still need to wait because we want to fail the offset commit for cases when
+                // tombstone records fail to be written to the secondary store. Note that while commitTransaction
+                // already waits for all records to be sent and ack'ed, in this case we do need to add an explicit
+                // blocking call. In case of ALOS, we wait for the same duration as `offset.commit.timeout.ms`
+                // and throw that exception which would allow the offset commit to fail.
+                if (isEOSEnabled) {
+                    secondaryWriteFuture.get();
+                } else {
+                    secondaryWriteFuture.get(offsetFlushTimeoutMs, TimeUnit.MILLISECONDS);
+                }
+            } catch (InterruptedException e) {
+                log.warn("{} Flush of tombstone offsets to secondary store interrupted, cancelling", this);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            } catch (ExecutionException e) {
+                log.error("{} Flush of tombstone offsets to secondary store threw an unexpected exception: ", this, e);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            } catch (TimeoutException e) {
+                log.error("{} Timed out waiting to flush offsets with tombstones to secondary storage ", this);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            } catch (Exception e) {
+                log.error("{} Got Exception when trying to flush tombstone offsets to secondary storage", this);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            }
+            Throwable writeError = secondaryStoreTombstoneWriteError.get();
+            if (writeError != null) {
+                FutureCallback<Void> failedWriteCallback = new FutureCallback<>(callback);
+                failedWriteCallback.onCompletion(writeError, null);
+                return failedWriteCallback;
+            }
+        }
+
         return primaryStore.set(values, (primaryWriteError, ignored) -> {
-            if (secondaryStore != null) {
+            // Secondary store writes have already happened for tombstone records
+            if (secondaryStore != null && !containsTombstones) {
                 if (primaryWriteError != null) {
-                    log.trace("Skipping offsets write to secondary store because primary write has failed", primaryWriteError);
+                    log.trace("Skipping offsets write to secondary store for non tombstone offsets because primary write has failed", primaryWriteError);
```

Comment:
This log line would look pretty confusing to most users, maybe we can just leave it as is since the `non tombstone offsets` bit isn't too relevant here?

Reply:
yeah, removed.